File: python/keystore_cli.py
#!/usr/bin/env python3
"""
Keystore CLI: read/write keystore storage.

Usage:
  python -m keystore_cli read [key]
  python -m keystore_cli write <key> <value>
  python -m keystore_cli delete <key>
  python -m keystore_cli init

Options:
  --path <file>    Keystore file path (default: ./.keystore.json)
  --password <pwd>  Password for encrypt/decrypt (or KEYSTORE_PASSWORD)
"""
import argparse
import os
import sys
from pathlib import Path

# Allow running from repo root or from python/
sys.path.insert(0, str(Path(__file__).resolve().parent))

from keystore import Keystore


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Keystore CLI - read/write keystore storage",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Commands:
  read [key]       Read one key or all keys
  write <key> <value>  Write one key
  delete <key>     Delete one key
  init             Create empty keystore file
        """,
    )
    parser.add_argument(
        "--path",
        default=os.environ.get("KEYSTORE_PATH", str(Path.cwd() / ".keystore.json")),
        help="Keystore file path",
    )
    parser.add_argument(
        "--password",
        default=os.environ.get("KEYSTORE_PASSWORD"),
        help="Encryption password",
    )
    parser.add_argument("command", nargs="?", choices=["read", "write", "delete", "init"], help="Command")
    parser.add_argument("args", nargs="*", help="Key and/or value")
    parsed = parser.parse_args()

    cmd = (parsed.command or "").lower()
    args = parsed.args or []
    file_path = parsed.path
    password = parsed.password

    ks = Keystore(file_path=file_path, password=password)

    if not cmd:
        parser.print_help()
        sys.exit(0)

    try:
        if cmd == "read":
            ks.read()
            key = args[0] if args else None
            if key:
                value = ks.get(key)
                if value is None:
                    print(f"Key not found: {key}", file=sys.stderr)
                    sys.exit(1)
                print(value)
            else:
                import json
                print(json.dumps(ks.get_all(), indent=2, ensure_ascii=False))

        elif cmd == "write":
            if len(args) < 2:
                print("Usage: write <key> <value>", file=sys.stderr)
                sys.exit(1)
            key, value = args[0], " ".join(args[1:]).strip().strip("'\"")
            ks.read()
            ks.set(key, value)
            ks.write()
            print(f"Written: {key}")

        elif cmd == "delete":
            if not args:
                print("Usage: delete <key>", file=sys.stderr)
                sys.exit(1)
            key = args[0]
            ks.read()
            data = ks.get_all()
            if key not in data:
                print(f"Key not found: {key}", file=sys.stderr)
                sys.exit(1)
            del data[key]
            Keystore.to_file(file_path, data, password)
            print(f"Deleted: {key}")

        elif cmd == "init":
            if os.path.isfile(file_path):
                print(f"File already exists: {file_path}", file=sys.stderr)
                sys.exit(1)
            Keystore.to_file(file_path, {}, password)
            print(f"Created: {file_path}")

    except Exception as e:
        print(str(e), file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: typescript/src/cli/keystore_cli.ts
#!/usr/bin/env node
/**
 * Keystore CLI: read/write keystore storage.
 *
 * Usage:
 *   npx ts-node src/cli/keystore_cli.ts read [key]     # read one key or all
 *   npx ts-node src/cli/keystore_cli.ts write <key> <value>  # write one key
 *   npx ts-node src/cli/keystore_cli.ts delete <key>   # remove one key
 *   npx ts-node src/cli/keystore_cli.ts init           # create empty keystore (optionally encrypted)
 *
 * Options:
 *   --path <file>    Keystore file path (default: ./.keystore.json)
 *   --password <pwd> Password for encrypt/decrypt (or set KEYSTORE_PASSWORD)
 */

import * as fs from 'fs';
import * as path from 'path';
import { Keystore } from '../keystore';

function getArg(name: string): string | undefined {
  const i = process.argv.indexOf(name);
  if (i === -1 || i >= process.argv.length - 1) return undefined;
  return process.argv[i + 1];
}

function hasFlag(name: string): boolean {
  return process.argv.includes(name);
}

async function main(): Promise<void> {
  const argv = process.argv.slice(2);
  const filePath = getArg('--path') ?? process.env.KEYSTORE_PATH ?? path.join(require('os').homedir(), '.agent_wallet', 'Keystore');
  const password = getArg('--password') ?? process.env.KEYSTORE_PASSWORD;

  const args: string[] = [];
  for (let i = 0; i < argv.length; i++) {
    if (argv[i] === '--path' || argv[i] === '--password') {
      i++;
      continue;
    }
    if (!argv[i].startsWith('--')) args.push(argv[i]);
  }
  const cmd = args[0]?.toLowerCase();

  const keystore = new Keystore({ filePath, password });

  if (!cmd || cmd === 'help' || hasFlag('--help') || hasFlag('-h')) {
    console.log(`
Keystore CLI - read/write keystore storage

  read [key]              Read one key or all keys
  write <key> <value>     Write one key (value can be quoted)
  delete <key>            Delete one key
  init                    Create empty keystore file

Options:
  --path <file>           Keystore file (default: ./.keystore.json)
  --password <pwd>        Encryption password (or KEYSTORE_PASSWORD)
`);
    return;
  }

  try {
    if (cmd === 'read') {
      const key = args[1];
      await keystore.read();
      if (key) {
        const value = await keystore.get(key);
        if (value === undefined) {
          console.error(`Key not found: ${key}`);
          process.exit(1);
        }
        console.log(value);
      } else {
        const all = await keystore.getAll();
        if (Object.keys(all).length === 0) {
          console.log('{}');
        } else {
          console.log(JSON.stringify(all, null, 2));
        }
      }
      return;
    }

    if (cmd === 'write') {
      const key = args[1];
      const value = args.slice(2).join(' ').replace(/^["']|["']$/g, '');
      if (!key || args.length < 3) {
        console.error('Usage: write <key> <value>');
        process.exit(1);
      }
      await keystore.read();
      await keystore.set(key, value);
      await keystore.write();
      console.log(`Written: ${key}`);
      return;
    }

    if (cmd === 'delete') {
      const key = args[1];
      if (!key) {
        console.error('Usage: delete <key>');
        process.exit(1);
      }
      await keystore.read();
      const all = await keystore.getAll();
      if (!(key in all)) {
        console.error(`Key not found: ${key}`);
        process.exit(1);
      }
      delete all[key];
      await Keystore.toFile(filePath, all, password);
      console.log(`Deleted: ${key}`);
      return;
    }

    if (cmd === 'init') {
      if (fs.existsSync(filePath)) {
        console.error(`File already exists: ${filePath}`);
        process.exit(1);
      }
      await Keystore.toFile(filePath, {}, password);
      console.log(`Created: ${filePath}`);
      return;
    }

    console.error(`Unknown command: ${cmd}. Use read, write, delete, init.`);
    process.exit(1);
  } catch (e) {
    console.error(e instanceof Error ? e.message : e);
    process.exit(1);
  }
}

main();
